cmd/skiff/top.go

```diff
@@ -66,31 +87,85 @@ func (h *FileHeap) Pop() interface{} {
 	return item
 }
 
+// getLayersByDiffID returns layer blob infos filtered by user-provided diffIDs
+// by looking up diffIDs and mapping to manifest layers
+func getLayersByDiffID(manifestLayers []types.BlobInfo, allDiffIDs []digest.Digest, filterDiffIDs []string) ([]types.BlobInfo, []digest.Digest, error) {
+	// If no filtering, return all layers with all their diffIDs
+	if len(filterDiffIDs) == 0 {
+		return manifestLayers, allDiffIDs, nil
+	}
+
+	// Filter layers by user-provided diffIDs
+	var filteredLayers []types.BlobInfo
+	var filteredDiffIDs []digest.Digest
+
+	// Map user diffIDs to layer indices
+	for _, userDiffID := range filterDiffIDs {
+		found := false
+		for i, configDiffID := range allDiffIDs {
+			// Match full diffID or prefix
+			if configDiffID.String() == userDiffID || strings.HasPrefix(configDiffID.Encoded(), userDiffID) {
+				if i < len(manifestLayers) {
+					filteredLayers = append(filteredLayers, manifestLayers[i])
+					filteredDiffIDs = append(filteredDiffIDs, configDiffID)
+					found = true
+					break
+				}
+			}
+		}
+		if !found {
+			return nil, nil, fmt.Errorf("diffID %s not found in image", userDiffID)
+		}
+	}
+
+	return filteredLayers, filteredDiffIDs, nil
+}
+
 // analyzeLayers fetches layers for a given image reference
 // reads the associated layer archives and lists file info
-func analyzeLayers(uri string, ctx context.Context, sysCtx *types.SystemContext) error {
+func analyzeLayers(ctx context.Context, sysCtx *types.SystemContext, uri string, layers []string, humanReadable bool) error {
+	// represents an image from any transport (docker://, containers-storage://, etc.)
 	img, _, err := skiff.ImageAndLayersFromURI(ctx, sysCtx, uri)
 	if err != nil {
 		return err
 	}
 
+	// image source that helps us fetch layers to eventually show files from the stream
 	imgSrc, err := img.Reference().NewImageSource(ctx, sysCtx)
 	if err != nil {
 		return err
 	}
 	defer imgSrc.Close()
 
-	h := &FileHeap{}
-	heap.Init(h)
+	// Get transport-specific layer blob infos
+	manifestLayers, err := skiff.BlobInfoFromImage(ctx, sysCtx, img)
+	if err != nil {
+		return fmt.Errorf("failed to get blob info from image: %w", err)
+	}
+
+	conf, err := img.OCIConfig(ctx)
+	allDiffIDs := []digest.Digest{}
 
-	files := make([]FileInfo, h.Len())
+	// only get them if the rootfs type is correct
+	if err == nil && conf != nil && conf.RootFS.Type == "layers" {
+		allDiffIDs = conf.RootFS.DiffIDs
+	}
+
+	// Check that manifestLayers and allDiffIDs have matching lengths
+	if len(manifestLayers) != len(allDiffIDs) {
+		return fmt.Errorf("manifestLayers (%d) and allDiffIDs (%d) length mismatch", len(manifestLayers), len(allDiffIDs))
+	}
 
-	layerInfos, err := skiff.BlobInfoFromImage(img, ctx, sysCtx)
+	// Get filtered layers and their diffIDs using the refactored function
```

Comment:
no need for that information
```suggestion
	// Get filtered layers and their diffIDs
```